src/drivers/imx/sdma.c

```diff
@@ -0,0 +1,716 @@
+	 * 1) Each descriptor corresponds to each of the
+	 *    config->elem_array elems; if we have more than
+	 *    MAX_DESCRIPTORS we bail outright.
+	 * 2) For each of them, store the host side (SDRAM side) as
+	 *    buf_addr and keep the FIFO address as a separate variable.
+	 *    Complain if this address changes between descriptors as we
+	 *    do not support this for now.
+	 * 3) Enable interrupts, set up transfer width, length of elem,
+	 *    wrap bit on the last descriptor, host side address, and
+	 *    finally the DONE bit so the SDMA can use the descriptors.
+	 * 4) The FIFO address will be stored in the context.
+	 * 5) Actually upload context now as we are inside DAI prepare.
+	 *    We have no other opportunity in the future.
+	 *
+	 * Note: Certain information such as HW event number or FIFO
+	 * watermark must be fetched during this function, however they
+	 * may not be actually available. Some way to allow for more
+	 * communication than usual between the components is needed.
+	 */
+	/* Check validity of config */
+	if (config->elem_array.count > SDMA_MAX_BDS) {
+		trace_sdma_error("sdma_set_config: Unable to handle %d descriptors",
+				 config->elem_array.count);
+		return -EINVAL;
+	}
+	if (config->elem_array.count <= 0) {
+		trace_sdma_error("sdma_set_config: Invalid descriptor count: %d",
+				 config->elem_array.count);
+		return -EINVAL;
+	}
+	bool src_may_change, dst_may_change;
+	switch (config->direction) {
+	case DMA_DIR_MEM_TO_DEV:
+		src_may_change = true; dst_may_change = false;
+		break;
+	case DMA_DIR_DEV_TO_MEM:
+		src_may_change = false; dst_may_change = true;
+		break;
+	case DMA_DIR_MEM_TO_MEM:
+		/* Fallthrough, TODO implement if desired, both are true */
+	default:
+		trace_sdma_error("sdma_set_config: Unsupported direction %d", config->direction);
+		return -EINVAL;
+	}
+	uint32_t fifo_paddr = 0;
+	for (int i = 0; i < config->elem_array.count; i++) {
+		if (!src_may_change) {
+			/* src is fifo */
+			if (i == 0)
+				fifo_paddr = config->elem_array.elems[0].src;
+			else if (fifo_paddr != config->elem_array.elems[i].src) {
+				trace_sdma_error("sdma_set_config: FIFO changes address!");
+				return -EINVAL;
+			}
+		}
+		if (!dst_may_change) {
+			/* dst is fifo */
+			if (i == 0)
+				fifo_paddr = config->elem_array.elems[0].dest;
+			else if (fifo_paddr != config->elem_array.elems[i].dest) {
+				trace_sdma_error("sdma_set_config: FIFO changes address!");
+				return -EINVAL;
+			}
+		}
+		if (config->elem_array.elems[i].size > SDMA_BD_MAX_COUNT) {
+			trace_sdma_error("sdma_set_config: elem transfers too much: %d bytes",
+					 config->elem_array.elems[i].size);
+			return -EINVAL;
+		}
+	}
+	/* Checks passed, we need to populate the buffer descriptors,
+	 * adjust count, configure context so the FIFO address is also
+	 * stored, upload context and have everything be ready.
+	 */
+
+	/* TODO fetch HW event number and proper watermark */
+	int hwevent = -1;
+	int watermark = 16;
```

Comment:
Magic numbers.

Reply:
Forgot to force push, these are already gone on my side. hwevent will come from handshake, watermark comes from dma config burst_elems.